core/src/client.rs

```diff
@@ -0,0 +1,231 @@
+	Other(failure::Error),
+}
+
+impl From<Error> for RpcError {
+	fn from(error: Error) -> Self {
+		RpcError::JsonRpcError(error)
+	}
+}
+
+/// The future retured by the client.
+pub struct RpcFuture {
+	recv: oneshot::Receiver<Result<Value, Error>>,
+}
+
+impl RpcFuture {
+	/// Creates a new `RpcFuture`.
+	pub fn new(recv: oneshot::Receiver<Result<Value, Error>>) -> Self {
+		RpcFuture {
+			recv
+		}
+	}
+}
+
+impl Future for RpcFuture {
+	type Item = Value;
+	type Error = RpcError;
+
+	fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
+		// TODO should timeout
+		match self.recv.poll() {
+			Ok(Async::Ready(Ok(value))) => Ok(Async::Ready(value)),
+			Ok(Async::Ready(Err(error))) => Err(RpcError::JsonRpcError(error)),
+			Ok(Async::NotReady) => Ok(Async::NotReady),
+			Err(error) => Err(RpcError::Other(error.into())),
+		}
+	}
+}
+
+/// A message sent to the `RpcClient`. This is public so that
+/// the derive crate can generate a client.
+pub struct RpcMessage {
+	/// The rpc method name.
+	pub method: String,
+	/// The rpc method parameters.
+	pub params: Params,
+	/// The oneshot channel to send the result of the rpc
+	/// call to.
+	pub sender: oneshot::Sender<Result<Value, Error>>,
+}
+
+/// A channel to a `RpcClient`.
+pub type RpcChannel = mpsc::Sender<RpcMessage>;
+
+/// The RpcClient handles sending and receiving asynchronous
+/// messages through an underlying transport.
+pub struct RpcClient<TSink, TStream> {
+	id: u64,
+	queue: HashMap<Id, oneshot::Sender<Result<Value, Error>>>,
+	sink: TSink,
+	stream: TStream,
+	ch: mpsc::Receiver<RpcMessage>
+}
+
+impl<TSink, TStream> RpcClient<TSink, TStream> {
+	/// Creates a new `RpcClient`.
+	pub fn new(
+		sink: TSink,
+		stream: TStream,
+		ch: mpsc::Receiver<RpcMessage>,
+	) -> Self {
+		RpcClient {
+			id: 0,
+			queue: HashMap::new(),
+			sink,
+			stream,
+			ch,
+		}
+	}
+
+	fn next_id(&mut self) -> Id {
+		let id = self.id;
+		self.id = id + 1;
+		Id::Num(id)
+	}
+}
+
+impl<TSink, TStream> Future for RpcClient<TSink, TStream>
+where
+	TSink: Sink<SinkItem=String, SinkError=RpcError>,
+	TStream: Stream<Item=String, Error=RpcError>,
+{
+	type Item = ();
+	type Error = RpcError;
+
+	fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
+		// Handle requests from the client.
+		loop {
+			let msg = match self.ch.poll() {
+				Ok(Async::Ready(Some(msg))) => msg,
+				Ok(Async::Ready(None)) => {
+					// When the channel is dropped we still need to finish
+					// outstanding requests.
+					break;
+				},
+				Ok(Async::NotReady) => break,
+				Err(()) => continue,
+			};
+			let id = self.next_id();
+			let request = Request::Single(Call::MethodCall(MethodCall {
+				jsonrpc: Some(Version::V2),
+				method: msg.method,
+				params: msg.params,
+				id: id.clone(),
+			}));
+			self.queue.insert(id, msg.sender);
+			let request_str = serde_json::to_string(&request)
+				.map_err(|error| RpcError::Other(error.into()))?;
+			self.sink.start_send(request_str)?;
```

Comment:
This is not handled correctly - the `start_send` may return `Ok(NotReady(Requested))` in case the stream is full. In such case the msg should be retried, see: https://docs.rs/futures/0.1.7/futures/enum.AsyncSink.html